### ISSUE TITLE
Add a `display_mermaid(diagram: str) function which rendered the provided diagram as a Mermaid diagram

### ISSUE THREAD

**Issue by afonit** [ipython/ipython#14319]
As mentioned by @dhirschfeld in https://github.com/jupyterlab/jupyterlab/issues/15725

With jupyterlab 4.1 coming out and supporting Mermaid.js diagrams, it would be convenient to display programmatically generated Mermaid diagrams in JupyterLab.

See this comment for the initial idea:
https://github.com/jupyterlab/jupyterlab/issues/15725#issue-2111206154

**Comment by Carreau** [comment]
Really I don't think we should, if something has a mermaid repr, then `display(obj)` should be sufficient. I don't think IPython should grow all those `display_*`

`display` have the include  and exclude keywords if you really want to customize which repr are sent, and in general the frontend should let you choose with a dropdown or anything else. 

The current `display_*` methods that exists are more a historical artifact than an API that should be mimicked.

**Comment by Carreau** [comment]
Also see [this talk](https://www.youtube.com/watch?v=GExKsQ-OU78) I did with Paul Ivanov  at JupyterCon arround 20 ish minutes you see a prototype where the frontend let you choise the repr you want, making `display_*` unecaesary anymore.

**Comment by dhirschfeld** [comment]
Sure, I actually agree with that - I was just trying to follow existing convention. All I care about is that I can view a str object as a mermaid diagram.

Maybe that just needs a `MermaidJS(DisplayObject)` class?

**Comment by dhirschfeld** [comment]
Maybe the issue title should be changed to:

#### Allow displaying a string variable as a Mermaid diagram

**Comment by Carreau** [comment]
You can register custom mimetype for any objects, even builtin types, or types you don't control methods of:

https://ipython.readthedocs.io/en/stable/config/integrating.html#formatters-for-third-party-types

https://ipython.readthedocs.io/en/stable/config/shell_mimerenderer.html

https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.formatters.html

That is if you want to display _all_ strings via mermaid. 

Otherwise yes, I believe you "just" need an object that define [`_repr_mimebundle`](https://ipython.readthedocs.io/en/stable/config/integrating.html#MyObject._repr_mimebundle_), which return a dict with the right informations, it does not need to inherit from DisplayObject.

**Comment by dhirschfeld** [comment]
Don't you need a `MermaidJS` mimetype that JupyterLab knows how to display?

From the current markdown implementation it looks like it's being converted to an inline svg:
```html
<div class="jp-RenderedMermaid" style="">
  <figure>
    <img src="data:image/svg+xml,%3Csvg%20<snip>">
  </figure>
</div>
```

**Comment by dhirschfeld** [comment]
NB: the `metadata` link points to [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html#metadata) which is a totally different thing right? (or am I just very confused!?)

https://github.com/ipython/ipython/blob/8b1204b6c4489c786ada294088a3acef7bd53e90/docs/source/config/integrating.rst?plain=1#L134

https://ipython.readthedocs.io/en/stable/config/integrating.html#more-powerful-methods

**Comment by dhirschfeld** [comment]
I can create a string variable, *representing* html, wrap it in the `display.HTML` class and have JupyterLab display a rendered view - e.g.

![image](https://github.com/ipython/ipython/assets/881019/d981ffa3-8c1f-4088-a010-499801d061e0)

I'd like to be able to do the same with a string object representing a mermaid diagram - import a class from `IPYthon.display`, wrap my string object and have JupyterLab display the rendered diagram:

![image](https://github.com/ipython/ipython/assets/881019/a7585528-3285-48a9-aee2-256af605b1a1)

Since not all strings are mermaid diagrams I think I need some custom class to wrap them and it would seem to make sense for that to live alongside the other rich repr classes in `IPython.display`?

**Comment by dhirschfeld** [comment]
> it would seem to make sense for that to live alongside the other rich-repr classes in `IPython.display`

...it doesn't have to though - I would be happy to define my own class:
```python
class MyMermaidJS:
    def _repr_mimebundle_(self, include=None, exclude=None) -> tuple[dict, dict]:
        ...
        return ???
```

...I'm just not too sure what I need to return for JupyterLab to understand how to display it as a rendered Mermaid diagram :thinking:

**Comment by dhirschfeld** [comment]
> Also see [this talk](https://www.youtube.com/watch?v=GExKsQ-OU78) I did with Paul Ivanov at JupyterCon arround 20 ish minutes you see a prototype where the frontend let you choose the repr you want, making `display_*` unecaesary anymore.

Yep - I know of your efforts in that area ( :heart: ) and I've thought for [a long time](https://github.com/jupyterlab/jupyterlab/issues/3038#issuecomment-351557769) that it would be very useful to have that capability.

Funnily enough, this came up for me again just the other week where the rich-repr of a `shapely` point is a rather useless dot:

![image](https://github.com/ipython/ipython/assets/881019/9ebcb7be-a6e1-43b5-a2a3-25c926eccb53)


...but since you can force the text-repr easily it wasn't a big deal for that use-case.

**Comment by Carreau** [comment]
> NB: the `metadata` link points to [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html#metadata) which is a totally different thing right? (or am I just very confused!?)

Gha... sphinx is resolving to the wrong thing indeed.

**Comment by Carreau** [comment]
Yes, JupyterLab would need to know about the mermaid mimetype, it would need a mimerender:

https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#mime-renderer-plugins

Here are a few renderer: https://github.com/jupyterlab/jupyter-renderers


It seem it's currently a provider, I'm not sure the difference between a provider and renderer, you woudl have to ask the JupyterLab team, but my guess is if there is a provider it's likely only a few line to add a render, you just need to decide the mimetipe (and file extension), but mermaid [already recommend `text/vnd.mermaid`, and `.mermaid`/`.mmd`](https://mermaid.js.org/ecosystem/integrations-create.html).

**Comment by Carreau** [comment]
There is already a notebook that should display a mermaid diagram:

https://github.com/jupyterlab/jupyterlab/blob/b2f36ff0a2a6c3c9f42ca4501593cb67e1c6a840/galata/test/jupyterlab/notebooks/mermaid_notebook.ipynb

it basically show you what `_repr_mime_` should return, except it returns _only_ mermaid, and you should also likely return `text/plain`, if you want a text repr for frontends that don't support mermaid

**Comment by dhirschfeld** [comment]
Cool, that seems to do the job! :rocket:

![image](https://github.com/ipython/ipython/assets/881019/f9794cc2-2507-418d-85cc-2ad6bf556d09)

**Comment by bollwyvl** [comment]
Yep, `.mmd` files and `text/vnd.mermaid` rendered by mermaid.js is only a _widely-implemented_ (e.g. on GitLab, GitHub, HackMD, and now, `nbconvert`, `jupyterlab`, and `notebook`) way to communicate, not a full multi-stakeholder-managed specification like `image/svg+xml` with many (albeit semi-incompatible) renderers. 

The `_repr_mimebundle_` approach for custom objects is the right way, today, and something that was _generating_ these displays dynamically could likely just as easily generate `text/vnd.graphviz` or other visual outputs.

Indeed, the argument could be made that things like `YoutubeVideo` are already a bridge too far, and even the venerable `LaTeX` renderers _famously_ doesn't work the way that the reference implementation does.

Some possible contribution targets for an actual `render_mermaid`/`Mermaid` for IPython:

- [ipylab](https://github.com/jtpio/ipylab)
- a new [jupyterlab-contrib](https://github.com/jupyterlab-contrib/) repo
- a new [jupyter-widgets](https://github.com/jupyter-widgets/) repo (actually _embracing_ the interactivity/hypertext features, instead of explicitly _avoiding_ them for security-at-rest, could be a really fun)

**Comment by dhirschfeld** [comment]
Would there be any interest in having this in `IPython.display` itself, alongside the `JSON`, `HTML`, `Markdown` and other `DisplayObject`s?

It's certainly where I'd first look for this functionality. (I'd be happy with just the `DisplayObject` subclass and not bothering with the `display_mermaid` function)

**Comment by MRYingLEE** [comment]
I am not familiar with technical detail. I thought Jupyter notebook was a web application and the ability to display any HTML (including the support of Mermaid diagram) was natural.  It seems not.

So may Jupyter provide a more general display_html function? If so, mermaid or any formats supported by web can be displayed. I understand there may be some security concerns.

Any ideas?

**Comment by cben** [comment]
@MRYingLEE [display_html](https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html) long exists:
```
from IPython.display import display_html
display_html('<p>A</p><script>alert("foo")</script><p>B</p>', raw=True)
```
there are also [HTML](https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.HTML) objects you can return, they implement [`_repr_html_` interface](https://ipython.readthedocs.io/en/stable/config/integrating.html#integrating-rich-display).  So "DIY" loading of Mermaid JS was always possible.  Security concerns limit sharing of such notebooks.
* E.g. packages such as https://github.com/vbalasu/mermaid-magic make it easy.  GitHub [can't run the JS](https://github.com/vbalasu/mermaid-magic/blob/main/mermaid-in-notebooks.ipynb) though [NBViewer can](https://nbviewer.org/github/vbalasu/mermaid-magic/blob/main/mermaid-in-notebooks.ipynb).

But now that Jupyter has built-in rendering of \```mermaid blocks in markdown cells — it'd be nice to have programmatic access to the _builtin_ mermaid rendering.

**Comment by cben** [comment]
- [x] ````IPython.display.Markdown('```mermaid\n...\n```')```` already renders.
- [x] as of jupyterlab 4.4.3, I see returning `text/vnd.mermaid` MIME gets not just rendered but enriched with `image/svg+xml` representation, saved into the .ipynb file :clap:  
  * This is great because none of [GitHub](https://github.com/orgs/community/discussions/163359)/GitLab render mermaid in .ipynb yet – not even in markdown.  
    Alas, they are also currently [pretty bad](https://github.com/jupyter/nbviewer/issues/1037) at rendering SVG, though [GitLab](https://gitlab.com/cben/sandbox/blob/6bdb770ff0a64e5187eed927b71ef866c056e70d/jupyter_notebook/mermaid_mime_type.ipynb) is almost fine 🥈.  
    UPDATE Sep 2025: [NBViewer](https://github.com/jupyter/nbviewer/issues/1075) now does render \```mermaid in markdown cells, as well as `text/vnd.mermaid` outputs, as well as SVG outputs 🥇.
  * This is much better than offloading diagram->image conversion to services like https://mermaid.ink or https://kroki.io/!  A few prior packages like https://github.com/ouhammmourachid/mermaid-py or https://github.com/sunhwan/jupyter-kroki-magic wrap such services conveniently, and the resulting images render well in any online notebook view.  
    But with `text/vnd.mermaid`, rendering is local & offline, no privacy leak, efficient (Mermaid inherently needs browser for layout, so these servers run headless browsers), and resulting .ipynb is self-contained for offline use too :tada:
